File: internal/domain/engine/engine.go
// Package engine содержит основной tick-цикл и обработчик событий.
// Engine — единственный мутатор runtime-состояния.
package engine

import (
	"context"
	"sync"
	"time"

	"github.com/TheCipherKeeper/cybercity-engine/internal/domain/models"
	"github.com/TheCipherKeeper/cybercity-engine/internal/domain/ports"
	"github.com/TheCipherKeeper/cybercity-engine/internal/domain/router"
	"github.com/TheCipherKeeper/cybercity-engine/internal/domain/state"
)

// Config — параметры, необходимые domain engine.
type Config struct {
	TickMs                int
	SnapshotIntervalTicks int
}

// HandlerResult — результат обработки события.
type HandlerResult struct {
	Children     []models.EventNode
	StateChanges []models.EventNode
}

// Handler — функция обработки события.
type Handler func(e *Engine, event models.EventNode) HandlerResult

// Engine — CyberCity simulation engine.
type Engine struct {
	Topology    *models.TopologyGraph
	Config      Config
	State       *state.StateManager
	Router      *router.EventRouter
	Store       ports.EventStore
	Snapshots   ports.SnapshotRepository
	Bus         ports.MessageBus
	Broadcaster ports.StateBroadcaster

	handlers  map[models.EventType]Handler
	commandCh chan models.EventNode
	pending   []models.EventNode
	running   bool
	mu        sync.Mutex
	stopCh    chan struct{}
}

// NewEngine создаёт новый экземпляр движка. Все зависимости передаются
// извне (dependency injection из application layer).
func NewEngine(
	topology *models.TopologyGraph,
	cfg Config,
	routerInst *router.EventRouter,
	store ports.EventStore,
	snapshots ports.SnapshotRepository,
	bus ports.MessageBus,
	broadcaster ports.StateBroadcaster,
) *Engine {
	if routerInst == nil {
		routerInst = router.NewEventRouter()
	}
	return &Engine{
		Topology:    topology,
		Config:      cfg,
		State:       state.NewStateManager(topology),
		Router:      routerInst,
		Store:       store,
		Snapshots:   snapshots,
		Bus:         bus,
		Broadcaster: broadcaster,
		handlers:    defaultHandlers(),
		commandCh:   make(chan models.EventNode, 256),
		pending:     make([]models.EventNode, 0),
		stopCh:      make(chan struct{}),
	}
}

// Start запускает tick-цикл движка. Блокирует до вызова Stop.
func (e *Engine) Start(ctx context.Context) error {
	e.mu.Lock()
	e.running = true
	e.mu.Unlock()

	ticker := time.NewTicker(time.Duration(e.Config.TickMs) * time.Millisecond)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			e.Stop()
			return ctx.Err()
		case <-e.stopCh:
			return nil
		case cmd := <-e.commandCh:
			e.mu.Lock()
			e.pending = append(e.pending, cmd)
			e.mu.Unlock()
		case <-ticker.C:
			if err := e.tick(ctx); err != nil {
				return err
			}
		}
	}
}

// Stop останавливает tick-цикл.
func (e *Engine) Stop() {
	e.mu.Lock()
	defer e.mu.Unlock()
	if !e.running {
		return
	}
	e.running = false
	close(e.stopCh)
}

// SubmitCommand отправляет команду в очередь движка.
func (e *Engine) SubmitCommand(command models.EventNode) error {
	select {
	case e.commandCh <- command:
	default:
	}
	return nil
}

// IsRunning возвращает true, если движок запущен.
func (e *Engine) IsRunning() bool {
	e.mu.Lock()
	defer e.mu.Unlock()
	return e.running
}

func (e *Engine) tick(ctx context.Context) error {
	e.State.IncrementTick()
	e.drainCommands()

	for len(e.pending) > 0 {
		event := e.pending[0]
		e.pending = e.pending[1:]
		e.ProcessEvent(ctx, &event)
	}

	if e.Broadcaster != nil && e.State.World.Tick%e.Config.SnapshotIntervalTicks == 0 {
		_ = e.Broadcaster.Broadcast(ctx, e.State.World)
	}

	if e.Config.SnapshotIntervalTicks > 0 && e.State.World.Tick%e.Config.SnapshotIntervalTicks == 0 {
		_ = e.Snapshots.Save(ctx, e.State.World.Tick, e.State.World)
	}

	return nil
}

func (e *Engine) drainCommands() {
	for {
		select {
		case cmd := <-e.commandCh:
			e.pending = append(e.pending, cmd)
		default:
			return
		}
	}
}

// ProcessEvent обрабатывает одно событие и распространяет его эффекты.
// Экспортируется для тестов.
func (e *Engine) ProcessEvent(ctx context.Context, event *models.EventNode) {
	_ = e.Store.Add(ctx, *event)

	handler := e.handlers[event.EventType]
	if handler == nil {
		event.Status = models.EventStatusSuppressed
		_ = e.Store.Add(ctx, *event)
		return
	}

	result := handler(e, *event)

	for _, child := range result.Children {
		_ = e.Store.Add(ctx, child)
		_ = e.Store.Link(ctx, event.EventID, child.EventID, models.EventEdgeKindPropagatedTo)
		e.pending = append(e.pending, child)
		if e.Bus != nil {
			_ = e.Bus.Publish(ctx, "city.events", child)
		}
	}

	for _, change := range result.StateChanges {
		_ = e.Store.Add(ctx, change)
		_ = e.Store.Link(ctx, event.EventID, change.EventID, models.EventEdgeKindTriggeredRule)
		if e.Bus != nil {
			_ = e.Bus.Publish(ctx, "city.events", change)
		}
		if change.TargetID != "" {
			sourceNode := e.Topology.Services[change.TargetID]
			if sourceNode != nil {
				propagated := e.Router.Propagate(change, sourceNode, e.Topology)
				e.pending = append(e.pending, propagated...)
			}
		}
	}

	event.Status = models.EventStatusProcessed
	_ = e.Store.Add(ctx, *event)
	if e.Bus != nil {
		_ = e.Bus.Publish(ctx, "city.events", *event)
	}
}

File: src/cybercity_engine/engine.py
"""Main engine loop and event processor.

The engine is the single authority over ``WorldState``. It consumes events
from the command queue and Redpanda, applies them through the state manager,
propagates them via the event router, and broadcasts resulting updates.
"""

from __future__ import annotations

import asyncio
from typing import Any

from .config import EngineConfig
from .events import EventGraph
from .models import EventNode, EventType, ServiceStatus, TopologyGraph
from .router import EventRouter
from .state import StateManager

__all__ = ["Engine"]


class Engine:
    """CyberCity simulation engine."""

    _handlers: dict[EventType, Any] = {}

    def __init__(
        self,
        topology: TopologyGraph,
        config: EngineConfig,
        router: EventRouter | None = None,
    ) -> None:
        self.topology = topology
        self.config = config
        self.state = StateManager(topology)
        self.router = router or EventRouter()
        self.event_graph = EventGraph()
        self._running = False
        self._command_queue: asyncio.Queue[EventNode] = asyncio.Queue()
        self._pending: list[EventNode] = []

    # ──────────────────────────────────────────────────────────────────
    # Public control
    # ──────────────────────────────────────────────────────────────────
    async def start(self) -> None:
        self._running = True
        while self._running:
            await self._tick()
            await asyncio.sleep(self.config.tick_ms / 1000.0)

    def stop(self) -> None:
        self._running = False

    async def submit_command(self, command: EventNode) -> None:
        """Enqueue a player/system command for processing."""
        await self._command_queue.put(command)

    # ──────────────────────────────────────────────────────────────────
    # Tick processing
    # ──────────────────────────────────────────────────────────────────
    async def _tick(self) -> None:
        self.state.increment_tick()

        # 1. Drain command queue
        commands = self._drain_commands()
        self._pending.extend(commands)

        # 2. Process pending events (FIFO for simplicity; later priority queue)
        while self._pending:
            event = self._pending.pop(0)
            await self._process_event(event)

        # 3. Background effects / recovery (placeholder)
        # TODO: background processes

    def _drain_commands(self) -> list[EventNode]:
        commands: list[EventNode] = []
        while not self._command_queue.empty():
            try:
                commands.append(self._command_queue.get_nowait())
            except asyncio.QueueEmpty:
                break
        return commands

    async def _process_event(self, event: EventNode) -> None:
        """Process a single event and propagate its effects."""
        self.event_graph.add(event)

        handler = self._handlers.get(event.event_type)
        if handler is None:
            event.status = "suppressed"
            return

        result = handler(self, event)
        if result is not None:
            child_events, state_changes = result
            for child in child_events:
                self.event_graph.add(child)
                self.event_graph.link(
                    event.event_id,
                    child.event_id,
                    "propagated_to",
                )
                self._pending.append(child)
            for change in state_changes:
                self.event_graph.add(change)
                self.event_graph.link(
                    event.event_id,
                    change.event_id,
                    "triggered_rule",
                )
                # State changes can also propagate through the graph.
                if change.target_id:
                    source_node = self.topology.services.get(change.target_id)
                    if source_node:
                        propagated = self.router.propagate(change, source_node, self.topology)
                        self._pending.extend(propagated)

        event.status = "processed"

    # ──────────────────────────────────────────────────────────────────
    # Event handlers
    # ──────────────────────────────────────────────────────────────────
    def _handle_heartbeat(self, event: EventNode) -> tuple[list[EventNode], list[EventNode]]:
        target_id = event.target_id
        if target_id is None:
            return [], []
        self.state.record_heartbeat(target_id, event.timestamp)
        return [], []

    def _handle_scan(self, event: EventNode) -> tuple[list[EventNode], list[EventNode]]:
        target_id = event.target_id
        if target_id is None:
            return [], []
        observer_id = event.source_id
        self.state.mark_seen_by(target_id, observer_id)
        source_node = self.topology.services.get(target_id)
        if source_node is None:
            return [], []
        propagated = self.router.propagate(event, source_node, self.topology)
        return propagated, []

    def _handle_attack(self, event: EventNode) -> tuple[list[EventNode], list[EventNode]]:
        target_id = event.target_id
        if target_id is None:
            return [], []
        success = event.payload.get("success", False)
        if not success:
            return [], []
        state_change = self.state.set_service_status(
            target_id, ServiceStatus.COMPROMISED, reason_event=event
        )
        return [], [state_change] if state_change else []

    def _handle_command(self, event: EventNode) -> tuple[list[EventNode], list[EventNode]]:
        action = event.payload.get("action")
        target_id = event.target_id
        if action == "ENABLE_BACKUP_POWER" and target_id:
            self.state.set_flag(target_id, "using_backup_power", True)
            change = self.state.set_service_status(
                target_id, ServiceStatus.UP, reason_event=event
            )
            return [], [change] if change else []
        if action == "ISOLATE_SERVICE" and target_id:
            change = self.state.set_service_status(
                target_id, ServiceStatus.MAINTENANCE, reason_event=event
            )
            return [], [change] if change else []
        return [], []

    def _handle_compromise(self, event: EventNode) -> tuple[list[EventNode], list[EventNode]]:
        target_id = event.target_id
        if target_id is None:
            return [], []
        change = self.state.set_service_status(
            target_id, ServiceStatus.COMPROMISED, reason_event=event
        )
        return [], [change] if change else []


# Bind handlers after class definition to avoid forward-reference issues.
Engine._handlers = {
    EventType.HEARTBEAT: Engine._handle_heartbeat,
    EventType.SCAN: Engine._handle_scan,
    EventType.ATTACK: Engine._handle_attack,
    EventType.COMPROMISE: Engine._handle_compromise,
    EventType.COMMAND: Engine._handle_command,
}
